cocos/renderer/backend/Device.h

```diff
@@ -113,11 +113,27 @@ class Device : public cocos2d::Ref
      */
     virtual void setFrameBufferOnly(bool frameBufferOnly) = 0;
 
+    /**
+     * Create an auto released Program.
+     * @param vertexShader Specifes this is a vertex shader source.
+     * @param fragmentShader Specifes this is a fragment shader source.
+     * @return A Program instance.
+     */
+    virtual Program* newProgram(const std::string& vertexShader, const std::string& fragmentShader) = 0;
+
+    /**
+     * Create an auto released engine built-in Program.
+     * @param type Specifes the program type.
+     * @return A Program instance.
+     */
+    virtual Program* createBuiltinProgram(ProgramType type);
+    
```

Comment:
Device do have to distinguish builtin or not. It just creates a program. Add `Program::getBuiltinProgram()` is a better idea.